Migrate to Files.new* creation of I/O streams during IDE boot

mdindoffer opened the pull request:
This change drops the usage of FileInputStream and FileOutputStream in classes that affect the boot up of the IDE in order to help relieve the pressure on the GC via circumventing the finalization.

I've measured the most notable offenders during the boot to reside in core, bootstrap and openide.filesystems modules:
![allocationhs_fis](https://user-images.githubusercontent.com/12681399/33808173-1e8ce318-dde2-11e7-8252-6b904bc60243.png)
![allocationhs_fos](https://user-images.githubusercontent.com/12681399/33808174-22e3e15a-dde2-11e7-9996-b182d1ad54d7.png)

However, the **vastly** more prominent and omnipresent case of finalizable input streams can be observed in the ZipFileInputStreams originating from ZipFile and JarFile objects:
![allocationhs_zipfis](https://user-images.githubusercontent.com/12681399/33808237-3fd9bffe-dde3-11e7-815e-f921351a3860.png)
As far as I know, there's no trivial solution to resolve those finalizations, but will look around for possible improvements there. 

As for the GC performance, I've repeatedly measured very small improvements in the GC invocations with default JVM settings:
Vanilla - http://gceasy.io/diamondgc-report.jsp?oTxnId_value=43107dc9-d949-4990-8585-684685aa4d02
PR - http://gceasy.io/diamondgc-report.jsp?oTxnId_value=3c5f83a2-05d6-4edf-9719-35d5fd14ff6c
The improvements are rather small, but please do note the measurements were taken on a new fairly powerful desktop PC with clean NB builds (zero projects imported, etc...). Therefore the gain might be bigger on lower end machines, where NB might be run (e.g. local schools?) or during real life usage scenarios. I encourage such tests, if in doubt.

geertjanw commented:
Thanks a lot. Someone with more knowledge about this area of NetBeans will need to review it, thank you very much for this important work you're doing.

jglick commented:
Generally a good idea. Just beware of `InvalidPathException`s, which would escape existing `catch (IOException` blocks, as the Jenkins project discovered in [JENKINS-43531](https://issues.jenkins-ci.org/browse/JENKINS-43531).

geertjanw commented:
Are there comments on this one, further thoughts, insights, or is it ready to be merged?

matthiasblaesing commented on openide.filesystems/src/org/openide/filesystems/JarFileSystem.java:
```diff
@@ -503,24 +503,16 @@ private InputStream getTemporaryInputStream(JarFile jf, JarEntry je, boolean for
         if (createContent || forceRecreate) {
             // JDK 1.3 contains bug #4336753
             //is = j.getInputStream (je);
```
The commented code and the comment can be removed.

mdindoffer replied:
Are you sure?
I left it there, because the code underneath still calls the 'getInputStream4336753()'.
I think this is really obsolete and the entire method can be deleted, but I wanted to make this PR focused on one thing.

matthiasblaesing replied:
Agreed - I misread the change. Sorry for the noise.

mdindoffer commented:
@matthiasblaesing AFAIK NIO is fully implemented on Linux and Windows and MacOs in OpenJDK/OracleJDK. However, the MacOs has had some problems with the WatchService performance. Not sure if it is resolved by now.

matthiasblaesing commented:
I think the change is good and we should pull it in. If there is no Veto raised till mid next week I'll pull this in.

@mdindoffer according to the records of the apache foundation you don't have an ICLA on file. It would be good if you would be willing to sign one, see here: https://www.apache.org/dev/new-committers-guide.html#cla. It is not strictly necessary, but makes it easier.

jtulach commented:
Unrelated, but: instead of relying on `WatchService`, there is `masterfs/macosx` that handles the listening in an efficient way, right @tzezula?

jtulach commented:
Assuming the tests pass, the changes look fine. +1

tzezula commented:
Yes the `masterfs/macosx` uses native `FSEventStream` to listen on file changes.
When the module is enabled the `FileObject` fires changes using this native code.

matthiasblaesing commented:
Please don't merge yet. This changeset causes unittests to fail, that worked correctly before. I'm looking currently into the boot strap module.

matthiasblaesing commented:
@mdindoffer you need to review this. You introduced failure in at least three of the eight touched modules. I had a quick look at the netbeans "Bootstrap" module and there are at least two problems:

- `Stamps#compareAndUpdateFile` catches `FileNotFoundException`, but NIO throws `NoSuchFileException`, this catch needs to be updated (line 439)
- `Stamps.Store#store` you switched from `new FileOutputStream(cacheFile, append);` to `Files.newOutputStream(cacheFile.toPath(), StandardOpenOption.APPEND);` you are missing the "file does not yet exists" case (`StandardOpenOption.CREATE`)

Please check the other modules. A general rereview of the changes is advised. I advise to check the unittests for all affected modules in master and compare these to the results with your changes applied.

vieiro commented:
Adding the "work-in-progress" tag to avoid accidental commits.
Please double-check we're not introducing more bugs when cleaning up code.
Thanks,
Antonio

mdindoffer commented:
Thanks everyone, I had a closer look at the change and fixed the test regressions. Apart from the two issues reported by @matthiasblaesing, there was a missing create option in NbLogging. The commit also revealed a bug in the SystemFileSystemTest which was failing due to leaked file streams. Fixed the test.
Do I need to squash?

vieiro commented:
Changing label "Code cleanup" with "enhancement", as I think it better reflects the intent.
@mdindoffer is this ready for a new review then?

mdindoffer commented:
Yes, ready for review.

matthiasblaesing commented:
I'm currently build the current state. There were two failing tests (not caused by this changeset) and I'll now check the state after I merged to master. If the build runs clean, I'll push it..

vieiro commented:
I reviewed it but had not time to run unit tests. Looks good.

matthiasblaesing commented:
@mdindoffer I came to the same conclusion as @vieiro and merged the changes. I squashed the commits and added two commits already mentioned.

Thank you working on this.
